vividus-studio-server/vividus-studio-plugin/src/main/java/org/vividus/studio/plugin/service/StepDefinitionResolver.java

```diff
@@ -113,9 +114,36 @@ public Stream<ResolvedStepDefinition> resolve(String documentIdentifier)
     }
 
     @Override
-    public void setStepDefinitions(Collection<StepDefinition> stepDefinitions)
+    public void refresh(Collection<StepDefinition> stepDefinitions)
     {
-        this.stepDefinitions.addAll(stepDefinitions);
+        List<StepDefinition> dynamicDefinitions = new ArrayList<>();
+        stepDefinitions.forEach(sd ->
+        {
+            if (sd.isDynamic())
+            {
+                dynamicDefinitions.add(sd);
+            }
+            else
+            {
+                this.staticStepDefinitions.add(sd);
+            }
+        });
+
+        if (!this.dynamicStepDefinitions.isEmpty())
+        {
+            Set<String> modules = dynamicDefinitions.stream()
+                    .collect(Collectors.groupingBy(StepDefinition::getModule, Collectors.toList()))
+                    .keySet();
+            this.dynamicStepDefinitions.removeIf(sd -> modules.contains(sd.getModule()));
+        }
+        this.dynamicStepDefinitions.addAll(dynamicDefinitions);
+
+        List<StepDefinition> allStepDefinitions = new ArrayList<>();
+        allStepDefinitions.addAll(this.staticStepDefinitions);
+        allStepDefinitions.addAll(this.dynamicStepDefinitions);
+
+        this.groupedStepDefinitions = Suppliers.memoize(() -> allStepDefinitions.stream()
+                .collect(Collectors.groupingBy(StepDefinition::getStepType, Collectors.toList())));
```

Comment:
why is Supplier needed now?

Reply:
for the same purpose as before, to group step definitions on demand, when we invoke refresh there are no warranties that the next call would be a call for step definitions